src/librustc_typeck/astconv.rs

```diff
@@ -327,146 +327,201 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
-                    "wrong number of {} arguments: expected {}{}, found {}",
-                    kind, quantifier, bound, provided,
-                ),
-                DiagnosticId::Error("E0107".into()),
-            );
-            for span in spans {
-                err.span_label(span, label.as_str());
-            }
-            err.emit();
+                    DiagnosticId::Error("E0107".into()),
+                );
+                for span in spans {
+                    err.span_label(span, label.as_str());
+                }
+                err.emit();
 
-            Err(true)
-        };
+                Err(())
+            };
 
         let mut arg_count_correct = Ok(());
         let mut unexpected_spans = vec![];
 
         if !infer_lifetimes || arg_counts.lifetimes > param_counts.lifetimes {
-            arg_count_correct = check_kind_count(
-                "lifetime",
-                param_counts.lifetimes,
+            arg_count_correct = check_lifetime_count(
                 param_counts.lifetimes,
                 arg_counts.lifetimes,
-                0,
                 &mut unexpected_spans,
                 explicit_late_bound == ExplicitLateBound::Yes,
-            )
-            .and(arg_count_correct);
+            );
         }
+
         // FIXME(const_generics:defaults)
-        if !infer_args || arg_counts.consts > param_counts.consts {
-            arg_count_correct = check_kind_count(
-                "const",
-                param_counts.consts,
-                param_counts.consts,
-                arg_counts.consts,
-                arg_counts.lifetimes + arg_counts.types,
-                &mut unexpected_spans,
-                false,
-            )
-            .and(arg_count_correct);
-        }
-        // Note that type errors are currently be emitted *after* const errors.
-        if !infer_args || arg_counts.types > param_counts.types - defaults.types - has_self as usize
+        let required_types = param_counts.types - param_counts.type_defaults - has_self as usize;
+        let permitted_types = param_counts.types - has_self as usize;
+        if !infer_args
+            && (arg_counts.consts != param_counts.consts
+                || arg_counts.types > permitted_types
+                || arg_counts.types < required_types)
         {
-            arg_count_correct = check_kind_count(
-                "type",
-                param_counts.types - defaults.types - has_self as usize,
-                param_counts.types - has_self as usize,
-                arg_counts.types,
-                arg_counts.lifetimes,
-                &mut unexpected_spans,
-                false,
-            )
-            .and(arg_count_correct);
+            // skip lifetime parameter definitions
+            let lifetimes = arg_counts.lifetimes;
+            // captures all the mismatches in parameter definitions
+            let matches = def.params.iter().zip(args.args.iter()).skip(lifetimes);
```

Comment:
you have to manually loop over `def.params` and `arg.args` here.

zip only returns something if both iterators return `Some`

Reply:
ah yea, I later iterate over just the excess arguments.

Reply:
I think so

That would be cleaner IMO :thinking:

Reply:
I'll try that out tmrw then